indexer/src/main.rs

```diff
@@ -31,27 +34,49 @@ fn run(home_dir: std::path::PathBuf, config: RunConfigArgs) -> Result<()> {
     };
 
     let system = actix::System::new();
+
+    let registry = Registry::new();
+    let server_handle = system.block_on(async {
+        if let Some(metrics_addr) = config.metrics_ip_port_address {
+            Some(run_metrics_server(metrics_addr, registry.clone()))
+        } else {
+            None
+        }
+    });
+
+    // TODO: refactor
     let block_res = system.block_on(async move {
         let indexer = near_indexer::Indexer::new(indexer_config).expect("Indexer::new()");
         let stream = indexer.streamer();
         let (view_client, _) = indexer.client_actors();
 
         // TODO: define buffer: usize const
         let (sender, receiver) = mpsc::channel::<CandidateData>(100);
+        let mut rabbit_publisher = rabbit_builder.build()?;
+        if let Some(_) = config.metrics_ip_port_address {
+            rabbit_publisher.enable_metrics(registry.clone())?;
+        }
+        let publisher_handle = rabbit_publisher.start()?;
 
-        let rabbit_publisher = rabbit_builder.build()?;
-
-        let block_listener = BlockListener::new(stream, sender, addresses_to_rollup_ids);
-        let receipt_validator = CandidatesValidator::new(view_client, receiver, rabbit_publisher);
+        let mut block_listener = BlockListener::new(stream, sender, addresses_to_rollup_ids);
+        let mut candidates_validator = CandidatesValidator::new(view_client, receiver, publisher_handle);
+        if let Some(_) = config.metrics_ip_port_address {
+            block_listener.enable_metrics(registry.clone())?;
+            candidates_validator.enable_metrics(registry.clone())?;
+        }
 
         let result = tokio::select! {
```

Comment:
I guess more accurate name would be `run` here. Those actually must work in parallel. I used `select` instead of join because as soon as one finished I want to exit. Tho   `system.run()?;` is used to wait until all routines finish so could be `join` reallly

Reply:
Make sense, if these are `run` methods. Don't you want to join for both to let the app to clean up the resources for all routines?